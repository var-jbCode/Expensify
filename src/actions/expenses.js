import { Children } from 'react'
import uuid from 'uuid'
import database from '../firebase/firebase'
//Regular
// Component calls action generators
// Action generator retunrs object 
// component dispatches object 
// redux store changes

// async Redux Actions
// Component calls action generators
// Action generator returns function
// component dispatches function 
// function runs


// add expense 
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

// remove expense
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})
// edit expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = {
            description,
            note,
            amount,
            createdAt
        }
        database.ref('expenses').push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }))
            })
    }
}

// SET_EXPENSES 
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = []
            snapshot.forEach(child => {
                expenses.push({
                    id: child.key,
                    ...child.val()
                })
            });
            dispatch(setExpenses(expenses))
        })
    }
}