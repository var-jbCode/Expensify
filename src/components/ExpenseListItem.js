import React from 'react';
import { NavLink } from 'react-router-dom'
import { connect } from 'react-redux'

export const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <NavLink to={`/edit/${id}`} activeClassName="is-active" exact={true}>
            <h3>{description}</h3>
        </NavLink>
        <p> Amount: {amount}, Created Date: {createdAt}</p>
    </div>
);



export default connect()(ExpenseListItem)