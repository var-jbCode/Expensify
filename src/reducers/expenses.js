//Expenses Reducer 

const expensesReducerDefaultstate = []

export default (state = expensesReducerDefaultstate, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ]
        case 'REMOVE_EXPENSE':
            return state.filter(expense => expense.id !== action.id)
        case 'SET_EXPENSES':
            return action.expenses
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense
                }
            })
        default:
            return state;
    }
}

