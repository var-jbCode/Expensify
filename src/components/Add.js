import React from 'react';
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { addExpense } from '../actions/expenses'

export class Add extends React.Component {
    onSubmit = (expense) => {
        props.onSubmit(expense)
        props.history.push('/')
    }
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    onSubmit: () => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(Add)

