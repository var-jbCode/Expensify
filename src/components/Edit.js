import React from 'react';
import { connect } from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm'

const Edit = (props) => {
    return (
        <div>
            <ExpenseForm
                expense={props.expense}
                onSubmit={(expense) => {
                    props.dispatch(editExpense(props.expense.id, expense))
                    props.history.push('/')
                }}
            />
            <button
                onClick={() => {
                    props.dispatch(startRemoveExpense({ id: props.expense.id }))
                    props.history.push('/')
                }
                }
            >Remove</button>
        </div>
    );
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}

export default connect(mapStateToProps)(Edit)
