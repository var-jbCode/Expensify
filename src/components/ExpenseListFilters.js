import React from 'react';
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'
import { DateRangePicker } from 'react-dates'

class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }
    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }))
    }
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input type="text"
                            value={this.props.filters.text}
                            onChange={(e) => {
                                this.props.dispatch(setTextFilter(e.target.value));
                            }} />
                    </div>
                    <div className="input-group__item">
                        <select
                            value={this.props.filters.sortBy}
                            onChange={(e) => {
                                if (e.target.value === 'date') { this.props.dispatch(sortByDate()) }
                                else if (e.target.value === 'amount') { this.props.dispatch(sortByAmount()) }
                            }}
                        >
                            <option value="date" >Date DESC</option>
                            <option value="amount" >Amount DESC</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => { false }}
                        />
                    </div>
                </div>



            </div>
        )

    }
}


const mapStatetoProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStatetoProps)(ExpenseListFilters)