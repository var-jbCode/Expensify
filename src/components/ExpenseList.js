import React from 'react';
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ) : (
                    props.expenses.map((expense) => (
                        <ExpenseListItem
                            key={expense.id}
                            id={expense.id}
                            description={expense.description}
                            amount={expense.amount}
                            createdAt={expense.createdAt}
                        />
                    ))
                )
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)

