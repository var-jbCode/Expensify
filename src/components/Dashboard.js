import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'


const Dashboard = () => (
    <div>
        <p>This is from I'm a rolling the beans Home </p>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default Dashboard