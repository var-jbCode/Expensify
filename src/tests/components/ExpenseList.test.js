import React from 'react';
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses'

test('should render Expense List with Expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render expens list with empty message', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})

