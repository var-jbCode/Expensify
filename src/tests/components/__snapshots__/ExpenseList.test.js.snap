// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense List with Expenses 1`] = `
<div>
  <Connect(ExpenseListItem)
    amount={195}
    createdAt={0}
    description="Gum"
    id={1}
    key="1"
  />
  <Connect(ExpenseListItem)
    amount={200000}
    createdAt={-345600000}
    description="rent"
    id={2}
    key="2"
  />
  <Connect(ExpenseListItem)
    amount={4500}
    createdAt={345600000}
    description="credit Card"
    id={3}
    key="3"
  />
</div>
`;

exports[`should render expens list with empty message 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;
